src/ModelContextProtocol/Server/McpServer.cs

```diff
@@ -318,23 +317,54 @@ private void SetToolsHandler(McpServerOptions options)
         {
             var originalListToolsHandler = listToolsHandler;
             var originalCallToolHandler = callToolHandler;
+            var hasOriginalTools = originalListToolsHandler is not null;
 
             // Synthesize the handlers, making sure a ToolsCapability is specified.
             listToolsHandler = async (request, cancellationToken) =>
             {
+                var progressToken = request.Params?.Meta?.ProgressToken;
+                int? totalProgress = hasOriginalTools ? null : tools.Count;
                 ListToolsResult result = new();
+                var resultTools = result.Tools;
+                    
+                async Task TryNotifyProgressAsync(int previousCount)
+                {
+                    var currentCount = resultTools.Count;
+                    if (progressToken is not null
+                        && currentCount > 0
+                        && currentCount != previousCount)
+                    {
```

Comment:
We don't need any of this anymore, do we? These conditions are all checked at the sole call site. And then once you remove all that, you can just inline the NotifyProgressAsync call and get rid of this method.

Reply:
And, with https://github.com/modelcontextprotocol/csharp-sdk/pull/177, this whole thing becomes moot, I think, as the server will no longer be in the business of aggregating across multiple calls. Each call from the client will result in at most a single call to a handler. That handler can choose to raise progress notifications if it likes.